### PR TITLE
CUDA

### PR DESCRIPTION
**Pull request by shadowpa0327**
## Codebase update
+ Update the definition of the grid. Previously, when we set the gridWidth to be N, there are only N-1 cells actually. This PR update the definition and the corresponding implementation such as display function and the implementation of the serial version.
## CUDA update
+ Add the CUDA kernel implementation, and successfully linking with CUDA runtime.

## Todo
+ Update the OpenMP version to align the new definition of the grid. cc @cm967413 
+ Update CUDA compilation to generate the hardware specific code by provideing the hardware information while trigger nvcc compiler.

### CODE REVIEW

File: gameOfLifeCUDA.cu [hunk 1]
```diff
@@ -1,25 +1,94 @@
...
+           2.0*prop.memoryClockRate*(prop.memoryBusWidth/8)/1.0e6);
+
     initGrid(mode, gridOne);
-    int iter = 0;
-    double elapseTime = 0.0;
+    
+
+    int size = arrayHeight * arrayWidth;
+    bool *d_gridOne, *d_gridTwo;
+    cudaError_t status;
```
Comment (f964966828) [inline gameOfLifeCUDA.cu]: What's the purpose of this variable?

File: gameOfLifeCUDA.cu [hunk 2]
```diff
@@ -1,25 +1,94 @@
...
+        size_t reqBlocksCount = ((gridWidth) * (gridHeight)) / threadCount;
+        updateCUDAKernel<<<reqBlocksCount, threadCount>>>(d_gridOne, d_gridTwo);
     }
+    cudaEventRecord(stop, 0);
+    cudaEventSynchronize(stop);
+    cudaEventElapsedTime(&elapseTime, start, stop);
+    cudaEventDestroy(start);
+    cudaEventDestroy(stop);
+    cudaMemcpy(gridOne, d_gridOne, size, cudaMemcpyDeviceToHost);
+    printGrid(gridOne);
```
Comment (f964966828) [inline gameOfLifeCUDA.cu]: Don't need to print gird each iteration?

File: gameOfLifeSerial.cpp [hunk 3]
```diff
@@ -40,6 +40,7 @@ double gameOfLifeSerial(bool* &gridOne, bool* &gridTwo, char mode){
             clearScreen();
         }
     }
+    printGrid(gridOne);
```
Comment (f964966828) [inline gameOfLifeSerial.cpp]: Same as above

File: gameOfLifeCUDA.cu (hunk 2 above)
Comment (shadowpa0327) [reply gameOfLifeCUDA.cu]: Sorry I try to print the last grid for debugging.
I will add a new commit to remove it

File: gameOfLifeCUDA.cu (hunk 1 above)
Comment (shadowpa0327) [reply gameOfLifeCUDA.cu]: It will be remove in the next commit

**Comment by f964966828** [comment]
LGTM !

**Comment by f964966828** [comment]
LGTM !

**Review by f964966828 — APPROVED** [review]
LGTM !

**Pull request merged**